pkg/controllers/dashboard/plugin/fscache.go

```diff
@@ -0,0 +1,253 @@
+			data, err := readFile(plugin.Endpoint + "/" + file)
+			if err != nil {
+				return err
+			}
+			path, err := filepathsecure.SecureJoin(FSCacheRootDir, filepath.Join(plugin.Name, plugin.Version, file))
+			if err != nil {
+				return err
+			}
+			if err := c.Save(data, path); err != nil {
+				logrus.Debugf("failed to cache plugin [Name: %s Version: %s] in filesystem [path: %s]", plugin.Name, plugin.Version, path)
+			}
+		}
+	}
+
+	return nil
+}
+
+// SyncWithIndex syncs up entries in the filesystem cache with the index's entries
+// Entries that aren't in the index, but present in the filesystem cache are deleted
+func (c FSCache) SyncWithIndex(index *SafeIndex, fsCacheFiles []string) error {
+	for _, file := range fsCacheFiles {
+		logrus.Debugf("syncing index with filesystem cache")
+		// Splits /{root}/{pluginName}/{pluginVersion}/* from a fs cache path
+		rel, _ := filepath.Rel(FSCacheRootDir, file)
+		s := strings.Split(rel, "/")
+		name := s[0]
+		version := s[1]
+		_, ok := index.Entries[name]
+		if !ok || index.Entries[name].Version != version {
+			err := c.Delete(name, version)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// Save takes in data and a path to save it in the filesystem cache
+func (c FSCache) Save(data []byte, path string) error {
+	logrus.Debugf("creating file [%s]", path)
+	if err := os.MkdirAll(filepath.Dir(path), os.ModePerm); err != nil {
+		return err
+	}
+	out, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+	out.Write(data)
+
+	return nil
+}
+
+// Delete takes in a plugin's name and version, and deletes its entry in the filesystem cache
+func (c FSCache) Delete(name, version string) error {
+	p, err := filepathsecure.SecureJoin(FSCacheRootDir, name)
+	if err != nil {
+		return err
+	}
+	err = osRemoveAll(p)
+	if err != nil {
+		err = fmt.Errorf("failed to delete entry [Name: %s Version: %s] from filesystem cache: %s", name, version, err.Error())
+		return err
+	}
+	logrus.Debugf("deleted plugin entry from cache [Name: %s Version: %s]", name, version)
+
+	return nil
+}
+
+// isCached takes in the name and version of a plugin and returns true if
+// it is cached (entry exists and files were fetched), returns false otherwise
+func (c FSCache) isCached(name, version string) (bool, error) {
+	path, err := filepathsecure.SecureJoin(FSCacheRootDir, filepath.Join(name, version))
+	if err != nil {
+		return false, err
+	}
+	_, err = osStat(path)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return false, nil
+		}
+		isEmpty, err := isDirEmpty(path)
+		if err != nil {
+			return false, err
+		}
+		if !isEmpty {
+			return true, nil
+		}
+
+		return false, err
+	}
+
+	if !errors.Is(err, os.ErrNotExist) {
+		isEmpty, err := isDirEmpty(path)
+		if err != nil {
+			return false, err
+		} else if !isEmpty {
+			return true, nil
+		}
+	} else if errors.Is(err, os.ErrNotExist) {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
```

Comment:
I think now this is redundant after the changes above.

```suggestion
```

Reply:
good catch, done